testing/src/main/java/org/jdbi/v3/testing/JdbiRule.java

```diff
@@ -13,26 +13,44 @@
  */
 package org.jdbi.v3.testing;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.UUID;
-
+import org.flywaydb.core.Flyway;
+import org.h2.jdbcx.JdbcConnectionPool;
 import org.jdbi.v3.core.Handle;
 import org.jdbi.v3.core.Jdbi;
 import org.jdbi.v3.core.spi.JdbiPlugin;
 import org.junit.rules.ExternalResource;
 
+import javax.sql.DataSource;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
 /**
  * JUnit {@code @Rule} to manage a Jdbi instance pointed to a managed database.
  */
 public abstract class JdbiRule extends ExternalResource {
 
+    private volatile DataSource dataSource;
     private Jdbi jdbi;
     private Handle handle;
     private boolean installPlugins;
+    private String[] migrationScriptPaths;
     private List<JdbiPlugin> plugins = new ArrayList<>();
 
-    protected abstract Jdbi createJdbi();
+    private final Object mutex = new Object();
+
+    protected abstract DataSource createDataSource();
+
+    private DataSource getDataSource() {
+        if (dataSource == null) {
```

Comment:
shame there isn't some single-value alternative to a ConcurrentHashMap with computeIfAbsent. AtomicReference only has such a method that takes a value, not one that takes a supplier of a value...

Reply:
Something like `Suppliers.memoize`?

Reply:
@stevenschlansker exactly